webapp-django/crashstats/crashstats/tests/test_utils.py

```diff
@@ -1,385 +1,411 @@
-
-        # A random string that does not match
-        input_str = 'somerandomstringthatdoesnotmatch'
-        assert not utils.find_crash_id(input_str)
-
-    def test_unicode_writer(self):
-        out = StringIO()
-        writer = utils.UnicodeWriter(out)
-        writer.writerow([
-            'abc',
-            u'\xe4\xc3',
-            123,
-            1.23,
-        ])
-        #NOTE: This probably needs to be a StringIO in Python 2 and a BytesIO in Python 3
-        # assuming UnicodeWriter even works in Python 3. I made change to binary_type for
-        # a simple fix until we run python 3 harness test in future.
-        result = out.getvalue()
-        assert isinstance(result, binary_type)
-        u_result = text_type(result, 'utf-8')
-        assert 'abc,' in u_result
-        assert u'\xe4\xc3,' in u_result
-        assert '123,' in u_result
-        assert '1.23' in u_result
-
-    def test_json_view_basic(self):
-        request = RequestFactory().get('/')
-
-        def func(request):
-            return {'one': 'One'}
-
-        func = utils.json_view(func)
-        response = func(request)
-        assert isinstance(response, HttpResponse)
-        assert json.loads(response.content) == {'one': 'One'}
-        assert response.status_code == 200
-
-    def test_json_view_indented(self):
-        request = RequestFactory().get('/?pretty=print')
-
-        def func(request):
-            return {'one': 'One'}
-
-        func = utils.json_view(func)
-        response = func(request)
-        assert isinstance(response, HttpResponse)
-        assert json.dumps({'one': 'One'}, indent=2) == response.content
-        assert response.status_code == 200
-
-    def test_json_view_already_httpresponse(self):
-        request = RequestFactory().get('/')
-
-        def func(request):
-            return HttpResponse('something')
-
-        func = utils.json_view(func)
-        response = func(request)
-        assert isinstance(response, HttpResponse)
-        assert response.content == 'something'
-        assert response.status_code == 200
-
-    def test_json_view_custom_status(self):
-        request = RequestFactory().get('/')
-
-        def func(request):
-            return {'one': 'One'}, 403
-
-        func = utils.json_view(func)
-        response = func(request)
-        assert isinstance(response, HttpResponse)
-        assert json.loads(response.content) == {'one': 'One'}
-        assert response.status_code == 403
+@pytest.mark.parametrize('version_string, expected', [
+    ('59.0', (59, 0, 0, 'zz')),
+    ('59.0.1', (59, 0, 1, 'zz')),
+
+    # alphas
+    ('59.0a1', (59, 0, 0, 'a1')),
+
+    # betas
+    ('59.0b1', (59, 0, 0, 'b1')),
+    ('59.0.1b1', (59, 0, 1, 'b1')),
+
+    # esr
+    ('59.0.1esr', (59, 0, 1, 'zz', 'esr')),
+
+    # junk data
+    ('49p', (-1)),
+])
+def test_parse_version(version_string, expected):
+    assert utils.parse_version(version_string) == expected
```

Comment:
I should add a test using `parse_version` as a key for sorting to make sure it sorts correctly now and to the end of time.